src/rgw/rgw_rados.cc

```diff
@@ -5875,25 +5875,42 @@ int RGWRados::select_new_bucket_location(RGWUserInfo& user_info, const string& z
 
   /* now check that tag exists within zonegroup */
   /* find placement rule. Hierarchy: request rule > user default rule > zonegroup default rule */
-  string rule = request_rule;
-  if (rule.empty()) {
-    rule = user_info.default_placement;
-    if (rule.empty())
-      rule = zonegroup.default_placement;
+  std::string rule;
+
+  if (!request_rule.empty()) {
+    if (zonegroup.has_placement_target(request_rule)) {
```

Comment:
note that these calls to `has_placement_target()` are duplicated with the final call to `zonegroup.placement_targets.find(rule);`. what do you think about tracking the selected rule by its iterator, instead of `std::string rule`? something like this:
```c++
  std::map<std::string, RGWZoneGroupPlacementTarget>::const_iterator titer;

  if (!request_rule.empty()) {
    titer = zonegroup.placement_targets.find(request_rule);
    if (titer == zonegroup.placement_targets.end()) {
      ldout(cct, 0) << "could not find requested placement id " << request_rule << " within zonegroup " << dendl;
      return -EINVAL;
    }
  } else if (!user_info.default_placement.empty()) {
    titer = zonegroup.placement_targets.find(user_info.default_placement);
    if (titer == zonegroup.placement_targets.end()) {
      ldout(cct, 0) << "could not find user default placement id " << user_info.default_placement << " within zonegroup " << dendl;
      return -EINVAL;
    }
  } ...
```
that avoids duplicating the final search through `zonegroup.placement_targets`, as well as the string copies into `std::string rule`. the new `RGWZoneGroup::has_placement_target()` also becomes unnecessary